[RFC] add audio renderer from chromium for better scaletempo

DorianRudolph opened the pull request:
I ported the [audio renderer algorithm](https://chromium.googlesource.com/chromium/chromium/+/51ed77e3f37a9a9b80d6d0a8259e84a8ca635259/media/filters/audio_renderer_algorithm.cc) from chromium.
I used the older version linked in #4418 because it seemed a little simpler than the most recent version.
In my opinion, chromium's algorithm sounds much better than `rubberband` and `scaletempo` (at least using the configurations I tried) and comparable to `atempo`.

The code seems to work fine, but I think it needs some more testing and refactoring.
I basically copied `af_rubberband.c` and `audio_renderer_algorithm.cc` from chromium including its dependencies into one file and hacked on it until it somehow worked. But I honestly don't really understand the code fully.

I have some questions:
- What should I do with frames that have `MP_NOPTS_VALUE`? This did not happen in my testing, but it seems to be handled by the other filters and I am not sure what PTS values I need to write for the corresponding output frames.
- Is `mp_aframe_get_planes` equal to the number of channels (for the planar format)?
- What happens after an EOF frame? It seems there can be further frames, but that seems odd to me. What behaviour is expected here? Right now I just rest the filter.
- Should I leave the code comments from chromium in?
- How can I best test all edge cases that can occur?
- What are good practices for how many samples to read and write?
- I think the last few samples will never be written to the output if not enough samples are there to `run_one_wsola_iteration`. Is this a problem?

This is my first open source contribution and I am looking forward to your feedback.

wm4 commented:
> What should I do with frames that have MP_NOPTS_VALUE? This did not happen in my testing, but it seems to be handled by the other filters and I am not sure what PTS values I need to write for the corresponding output frames.

This is rare and only happens with broken streams/filters/demuxers. It normally shouldn't happen at all, because the player wouldn't be able to show playback time or synchronize audio/video. Still, when it happens, it shouldn't produce random nonsense timestamps by treating MP_NOPTS_VALUE as a numeric value (e.g. adding small values to it etc.). The policy is that MP_NOPTS_VALUE timestamps are simply passed through.

> Is mp_aframe_get_planes equal to the number of channels (for the planar format)?

Yes. For non-planar formats, it's always 1.

> What happens after an EOF frame? It seems there can be further frames, but that seems odd to me. What behaviour is expected here? Right now I just rest the filter.

Yes, after EOF has been handled, a filter is supposed to go back into the initial state, like after reset(). That means you can just call reset internally once you have returned EOF. In theory, it can happen that the decoder restarts sending new frames, but this is extremely obscure and I'm currently not aware of a way to intentionally trigger this situation. Maybe I should think about a mechanism that does this.

> Should I leave the code comments from chromium in?

Not sure. Is there any specific reason to trigger them? If they reference internal issue numbers or so, it'd probably be better to replace them. Same for comments that matter only in context of chromium, instead of e.g. the audio algorithm.

> How can I best test all edge cases that can occur?

Only manually. This is a problem for me as well. Can you list what exactly you'd like to have tested? Maybe it wouldn't be so hard to create some sort of unittest for such filters.

> What are good practices for how many samples to read and write?

That's a good question. Currently, it's fairly random. Generally, a filter must accept a full input frame, but can produce an arbitrary number of output frames with arbitrary sample counts. f_swresample.c makes it configurable with ``--audio-resample-max-output-size``. Having this as option doesn't have much value (other than debugging). But limiting the output size to something small makes sense, so speed changes react faster. This might matter for A/V synchronization, or corner cases where frame sizes are extremely large.

I actually planned to add a way for filters to communicate "suggested" output/input frames, but it never materialized, because I doubted the value of doing that. Maybe later.

For now, I can only suggest to keep the code as flexible, and to limit output frame sizes to something arbitrary, maybe 50ms?

> I think the last few samples will never be written to the output if not enough samples are there to run_one_wsola_iteration. Is this a problem?

I'm not sure. The other filters always try to flush out all buffered data on EOF. Anyway, the most important part is that your filter doesn't get stuck on EOF (because no new input is coming, nor should you request new input).

---

Sorry, I haven't taken a close look at your code. I'll try to review it at some later point. But if you came this far, you're probably done with the hard parts. Good job somehow getting through the mpv filter code, you're the first person to do so.

DorianRudolph commented:
Thanks for the answers.

There are only comments regarding the implementation of the audio algorithm, so I suppose I will leave them in.

Regarding test cases, I think it would be nice to test everything that can occur like speed changes, reset, destroy, time jumps (when seeking), different frame sizes, EOF, and MP_NOPTS_VALUE. Though it is not clear to me, how to verify behaviour other than not crashing and that the audio sounds fine.

For EOF, I think will change the behavior such that it sends "zero" samples to the algorithm until it has enough data to flush all remaining input.

Besides that, I am working on refactoring the code to make it more "c-like" and avoid malloc/free calls on every frame.

DorianRudolph commented:
I just refactored the code. I think it can be merged unless there are some issues. I tested it with a few files and youtube videos and it seems to work fine.

megapro17 commented:
Do you have compiled build?

DorianRudolph commented:
I am not sure how to properly build mpv, but here is the mpv file from my build directory:
[mpv.zip](https://github.com/mpv-player/mpv/files/4874976/mpv.zip)
Security wise, its probably better if you just build it yourself...

avih commented on DOCS/man/af.rst:
```diff
@@ -193,6 +193,17 @@ Available filters are:
         If you want to go up or down by semi-tones, use 1.059463094352953 and
         0.9438743126816935
 
+``chromium``
```
Do we actually want to call it "chromium"? what if chromium changes its algorithm tomorrow?

DorianRudolph replied:
That's a good point. I'm fine with anything. Maybe `scaletempo2` because it is a very similar algorithm to that one?

avih replied:
> Maybe `scaletempo2`

Funnily enough, I thought of exactly this name too, though I wasn't sure if I should mention it as it's a bit weird :)

megapro17 replied:
i think better stay with chromium because it very well made

haasn replied:
I would call it `wsola`, because that seems to be what the algorithm itself is called in the literature. If need be you can mention that chromium uses it in the documentation for that option.

megapro17 replied:
No it's confusing there's a lot WSOLA algorithms that sounds worse than chromium

haasn replied:
Oh, I see. Well then, I guess it's fine. I just think it's a weird name for an audio algorithm, after all. "chromium" could mean any number of things. It doesn't really communicate what this exact filter is doing at all.

I'd be fine with the naming if it was in the context of an option like ``--audio-pitch-shift-algorithm=chromium`, but as it stands it's basically a stand-alone audio filter, `--af=chromium`. Maybe it could be `wsola-chromium` to distinguish this particular WSOLA variant from others?

megapro17 replied:
`wsola-chromium` sounds good

TiGR replied:
`scaletempo-ng`?

avih replied:
It's subjective, but in retrospect, i like `scaletempo2` the most, because It's an internal filter which is an alternative to scaletempo and has the same goal.

Maybe off topic, doesn't ffmpeg have such things with its filters?

DorianRudolph replied:
@avih Ok, then I will change it to `scaletempo2`.
ffmpeg also has this, but apparently it can't be used due to a missing API (https://github.com/mpv-player/mpv/issues/4418#issuecomment-636095336).

alopatindev replied:
Does original scaltempo implementation still have any useful use case now? How about replacing existing scaletempo and keeping the same name?

garoto replied:
I use scaletempo instead of rubberband currently. Perhaps in the event that this patch is merged and works as advertised, then perhaps a filter rename could be a thing.

haasn commented:
Is it possible you could add an option to do pitch shifting without changing the tempo, sort of like how `--af=rubberband` has the `pitch-scale` option, and `--af=scaletempo` has the `speed=pitch` option?

Sometimes I want to use these algorithms not for pitch-correction when changing the playback speed but to do pitch shifting on its own.

DorianRudolph commented:
@haasn Yes, I think I can add the same the same `speed=pitch` option. I suppose I can also expose `ola_window_size_ms` and `wsola_search_interval_ms` constants as parameters, although I am not sure what sensible ranges are.

In the above change, I just removed an unused variable. The pipeline failure seems to be unrelated?

DorianRudolph commented:
I renamed it and added parameters for `ola_window_size_ms` and `wsola_search_interval_ms`. However, changing them only seems to make the quality worse.

@haasn I thought it would be easy to do the scaletempo thing with `speed=pitch` but I am not sure if I understand what that even does.
I would expect `mpv <file> --af=scaletempo=speed=pitch --speed=2` increase the pitch by a factor of 2. I tested this with a 200hz sine wave and it does not do anything (for a normal video, voices just sound "roboty" and it also desyncs). Rubberband works as expected (although it introduces some reverb(?)).

Chromium does not have a pitch change feature. One way to implement it (and I have not read up on the proper theory papers for this) is to resample the audio and then change its speed.

To change the pitch by 1.5, you can do `--af="asetrate=48000*1.5,aresample=48000,atempo=1/1.5"`. But rubberband sounds a little bit better in my option.
I also tried `--af="asetrate=48000*1.5,aresample=48000,scaletempo2" --speed=0.666666`, but that slows down and pitches up at the same time (maybe some issue with the PTS stuff?).

DorianRudolph commented:
Is there anything that needs to be done before this can be merged?

wm4 commented:
In can't say much. I don't know the algorithm and its implementation, and I find it hard to tell whether the data flow on the mpv side is correct. (I've made more than one mistake when changing the other filters to the new filter API. Suggestions how to make it simpler are welcome. m:n filter dataflow is always messy, I guess.)

I'm noticing that there's a lot of audio data memcpy and memmove.

Would you agree that it's a good idea to split the filter into two files, one that has the mpv glue code, and one that has the chromium code?

Also, all commits should be squashed into one. There are some coding style issues. I'm also not sure how helpful muting audio outside of a min/max speed range is.

DorianRudolph commented:
I split up the internals as you suggested.

Regarding memcpy and memmove, the basic flow of audio data is as follows: mp_aframe -> input buffer -> WSOLA window -> output buffer -> mp_aframe

I don't think this is too much. I could probably get rid of the memmoves by using a ringbuffer instead of moving the data to the beginning. But I don't think this would be worth the additional work and code complexity. I profiled mpv with callgrind (first time using the tool, so not 100% sure I did it correctly) and scaletempo2 uses like 0.15% of the cycles whereas the subsequent resampling uses around 3% (mp3 playback).

The muting is mainly there because it was in chromium, but I think it might be helpful so I left it in. If you speed up or slow down playback too much, the audio is basically unintelligble, so it might be nice to have the option to have it muted automatically (it can be disabled).

wm4 commented:
There's TOOLS/uncrustify.cfg, but it doesn't always produce good results. Most of the weird things are in the chromium-derived code, and might come from the google code style. But whatever.

By memcpy/memmove I meant mostly that there are many functions using it, not necessarily that audio data is copied too much.

avih commented:
The `--speed` docs state:
> If --audio-pitch-correction (on by default) is used, playing with a speed higher than normal automatically inserts the scaletempo audio filter.

Is it possible that it will use scaletempo2 in such case instead of scaletempo?

DorianRudolph commented:
This feature seems to be hardcoded to scaletempo. Note that scaletempo2 does not change the audio at all for speed 1. So besides a small overhead, there should be no downside to having it permanently enabled.

avih commented:
> there should be no downside to having it permanently enabled

Apparently, with it permanently enabled (tested with consecutive invocations of the same mp3 file and `--af-add=scaletempo2`), after changing speed, quitting, and opening the same file again such that it's resumed, the `speed` property _value_ is resumed correctly as well (as expected), but actual playback speed (i.e. by listening) is as if `--speed=1` was used.

If not using `--af-add=scaletempo2` then both the property value and the actual playback speed match and resumed correctly.

So it seems that enabling it permanently creates issues while resuming from watch-later if the speed was modified.

avih commented:
Actually, it's easy to reproduce an issue. Just run `mpv --no-config whatever.mp3 --video=no --speed=2 --af=scaletempo2` and notice that it plays at normal speed. If you remove the `--af=scaletempo2` option then it plays at double speed.

Interestingly, if the file has cover art and `--no-video` is NOT used (i.e. a window opens with the cover art), then it does play at double speed even with scaletempo2.

DorianRudolph commented:
The bug should be fixed in https://github.com/mpv-player/mpv/pull/7950
